### ISSUE TITLE
Add other variants to std::io::ErrorKind enum

### ISSUE THREAD

**Issue by Alphare** [rust-lang/rust#64314]
I was trying to rewrite Python code in Rust and I need to match against what could be `NotADirectory` (`ENOTDIR`) and saw that `ErrorKind`'s variants do not include many `errno`s present in `libc` - to speak only of the Unix platform - including the one I need.

Is there a reason why those variants were not implemented? Could this be my first contribution to Rust?

**Comment by Mark-Simulacrum** [comment]
The docs do say "This list is intended to grow over time and it is not recommended to exhaustively match against it." currently (and we enforce non exhaustive matching via an attribute); I think our hands technically aren't tied here but we'll want to be very cautious.

@rust-lang/libs -- does anyone have thoughts on growing `io::ErrorKind` to have more variants? I don't recall us doing so in the last several years...

To address your specific concern, you can use `raw_os_error()` on `io::Error` and compare with libc's ENOTDIR as needed.

**Comment by alexcrichton** [comment]
I agree that it's been so long since we've added one that we should probably do so carefully (aka use crater) but I think in general it's fine to add more.

**Comment by SimonSapin** [comment]
The enum has `#[non_exhaustive]` so I’m not too worried a priori.

**Comment by Alphare** [comment]
Is this something I can reasonably tackle as a first contribution?

**Comment by SimonSapin** [comment]
Add variants sounds fine, but when would `NotADirectory` be returned? If it’s in existing APIs, what do they return today?

**Comment by Alphare** [comment]
@SimonSapin I think one case where it can happen is when you try a filesystem operation on `a/b` where `a` is not a directory. As of now it falls into the `Other` variant, but the match clause should translate [this constant](http://rust-lang.github.io/regex/src/libc/unix/notbsd/mod.rs.html#388) into `NotADirectory`.

**Comment by fmease** [comment]
Nowadays, `io::ErrorKind` has 41 variants (incl. `Other`) which includes `NotADirectory` mentioned in the issue description. 3 are currently unstable. See <https://doc.rust-lang.org/std/io/enum.ErrorKind.html>.

I consider this issue fixed. For context, there was a huge push towards extending `ErrorKind` a two, three (?) years ago which lead to this change. Can't link, I'm on mobile rn.
